pipeline-runner/R/qc-4-filter_gene_umi_outlier.R

```diff
@@ -25,85 +25,80 @@
-      if (exists("auto", where = config)) {
-        if (as.logical(toupper(config$auto))) {
-          p.level <- min(0.001, 1 / ncol(scdata))
-        }
-      }
-
-      # We regress the molecules vs the genes. This information are stored in nCount_RNA and nFeature_RNA respectively
-      df <- data.frame(molecules = sample_subset$nCount_RNA, genes = sample_subset$nFeature_RNA)
-      # We take log10 following the plot from the mock-up
-      df <- log10(df)
-      # Rename the rows to be able to identify the valid cells
-      rownames(df) <- colnames(sample_subset)
-      df <- df[order(df$molecules, decreasing = FALSE), ]
-      m <- MASS::rlm(genes ~ molecules, data = df)
-      # Get the interval based on p.level paramter
-      suppressWarnings(pb <- data.frame(predict(m,
-        interval = "prediction",
-        level = 1 - p.level, type = "response"
-      )))
-      # Define the outliers those that are below the lower confidence band and above the upper one.
-      outliers <- rownames(df)[df$genes > pb$upr | df$genes < pb$lwr]
-
-      # Keep the ones that are not outlier
-      # Because we have subseted the object before creating the dataframes, and just selected the outlier values from that df
-      # In the end we wil only remove values that are in the desired sample.
-      scdata.filtered <- subset_safe(scdata, colnames(scdata)[!colnames(scdata) %in% outliers])
-      # Similarly, when creating the plot from the df, we will just be using the values for the required sample.
-      plot1_data <- unname(purrr::map2(df$molecules, df$genes, function(x, y) {
-        c("log_molecules" = x, "log_genes" = y)
-      }))
-      plot1_data <- purrr::map2(plot1_data, unname(pb$lwr), function(x, y) {
-        append(x, c("lower_cutoff" = y))
-      })
-      plot1_data <- purrr::map2(plot1_data, unname(pb$upr), function(x, y) {
-        append(x, c("upper_cutoff" = y))
-      })
-      # have downsampling done
-
-      # Downsample plotData
-      num_cells_to_downsample <- downsample_plotdata(ncol(sample_subset), num_cells_to_downsample)
-
-      set.seed(123)
-      cells_position_to_keep <- sample(1:ncol(sample_subset), num_cells_to_downsample, replace = FALSE)
-      cells_position_to_keep <- sort(cells_position_to_keep)
-      plot1_data <- plot1_data[cells_position_to_keep]
-    }
-  } else {
-    scdata.filtered <- scdata
-    plot1_data <- list()
-  }
+  type <- config$filterSettings$regressionType
 
-  config$filterSettings$regressionTypeSettings[[config$filterSettings$regressionType]][["p.level"]] <- p.level
+  # get p.level and update in config
+  # defaults from "gene.vs.molecule.cell.filter" in pagoda2
+  if (!is.null(config$auto) && config$auto)
+    p.level <- min(0.001, 1 / ncol(scdata))
+  else
+    p.level <- config$filterSettings$regressionTypeSettings[[type]]$p.level
```

Comment:
Perhaps if auto is null (should never be) we should still do auto value? Assuming here that if things are so screwed up that auto is null we probably won't have a p.level

Reply:
Just a detail, really.

Reply:
If we don't have a `p.level` maybe we don't have a `regressionTypeSettings` or a `filterSettings` either and we need to check those before we overwrite the previous config `p.level`? If we are going to take this type of extreme defense (not opposed to it per se) I think it needs to be a larger discussion as the assumption to date has been that we receive well formed requests from the API.